### ISSUE TITLE
Server crash seemingly related to death of a deer not halting AI processing of the entity

### ISSUE THREAD

**Issue by Cervator** [MovingBlocks/Terasology#3623]
Noticed in the server log for the March play test just now. I was not online but my theory is somebody killed a deer then something about cleanup or parallel processing led to a server crash:

```
15:55:25.808 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
15:56:20.719 [main] INFO  o.t.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:56:20.720 [main] INFO  o.t.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:56:26.061 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
15:57:13.254 [main] ERROR o.t.e.e.internal.PojoEntityManager - Saving a component (class org.terasology.behaviors.components.FleeComponent) that doesn't belong to this entity 47898
15:57:13.255 [main] INFO  o.t.n.internal.NetworkSystemImpl - Component class org.terasology.logic.characters.CharacterMovementComponent removed from EntityRef{id = 47898, netId = 7956, prefab = 'WildAnimals:deer'}
15:57:13.255 [main] INFO  o.t.n.internal.NetworkSystemImpl - Component class org.terasology.wildAnimals.Death.DestroyAtAnimationEndComponent added to EntityRef{id = 47898, netId = 7956, prefab = 'WildAnimals:deer'}
15:57:13.255 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
15:57:13.255 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
15:57:13.255 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
15:57:13.255 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
15:57:13.255 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
java.util.ConcurrentModificationException: null
        at java.util.HashMap$HashIterator.nextNode(HashMap.java:1429)
        at java.util.HashMap$EntryIterator.next(HashMap.java:1463)
        at java.util.HashMap$EntryIterator.next(HashMap.java:1461)
        at org.terasology.logic.characters.ServerCharacterPredictionSystem.update(ServerCharacterPredictionSystem.java:194)
        at org.terasology.engine.modes.StateIngame.update(StateIngame.java:171)
        at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:458)
        at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:421)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:397)
        at org.terasology.engine.Terasology.main(Terasology.java:145)
15:57:13.256 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
15:57:13.810 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
15:57:13.812 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
15:57:13.823 [main] INFO  o.t.n.internal.NetworkSystemImpl - Client disconnected: Waterpicker
15:57:13.827 [main] INFO  o.t.n.internal.NetworkSystemImpl - Network shutdown
15:57:18.598 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextNode(HashMap.java:1429)
        at java.util.HashMap$EntryIterator.next(HashMap.java:1463)
        at java.util.HashMap$EntryIterator.next(HashMap.java:1461)
        at org.terasology.logic.characters.ServerCharacterPredictionSystem.update(ServerCharacterPredictionSystem.java:194)
        at org.terasology.engine.modes.StateIngame.update(StateIngame.java:171)
        at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:458)
        at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:421)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:397)
        at org.terasology.engine.Terasology.main(Terasology.java:145)
For more details, see the log files in /opt/terasology/server/logs/2019-03-02_12-00-30_2.2.0_78
```

Specifically the server was running Alpha 12 with JS + ShatteredPlanes + WildAnimalsGenome and I had previously activated deer mating once to make a baby deer. Crash occurred way later.

**Comment by Cervator** [comment]
Just happened again, coincidentally almost exactly a year later. JoshariasSurvival in ShatteredPlanes with WildAnimalsGenome active. Oddly: nobody else was online, so I'm not even sure how a deer could have died?

**Comment by skaldarnar** [comment]
I think this was fixed by #4278 (error during character movement prediction caused by killing mobs).
